src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.ChecksumToSolutionCache.cs

```diff
@@ -0,0 +1,171 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    internal sealed partial class RemoteWorkspace
+    {
+        /// <summary>
+        /// Cache of recently requested solution snapshots.  This always stores the last snapshot requested, and also
+        /// stores any requested solution snapshot that is currently in-use.  This allows concurrent calls to come in
+        /// and see/reuse in-flight solution snapshots being used by other requests.
+        /// </summary>
+        private sealed partial class ChecksumToSolutionCache
+        {
+            /// <summary>
+            /// Pointer to <see cref="RemoteWorkspace._gate"/>.
+            /// </summary>
+            private readonly SemaphoreSlim _gate;
+
+            /// <summary>
+            /// The last checksum and solution requested by a service. This effectively adds an additional in-flight
+            /// count to one of the items in <see cref="_solutionChecksumToSolution"/> ensuring that the very last
+            /// solution requested is kept alive by us, even if there are no active requests currently in progress for
+            /// that solution.  That way if we have two non-concurrent requests for that same solution, with no
+            /// intervening updates, we can cache and keep the solution around instead of having to recompute it.
+            /// </summary>
+            private Checksum? _lastRequestedChecksum;
+            private SolutionAndInFlightCount? _lastRequestedSolution;
+
+            /// <summary>
+            /// Mapping from solution-checksum to the solution computed for it.  This is used so that we can hold a
+            /// solution around as long as the checksum for it is being used in service of some feature operation (e.g.
+            /// classification).  As long as we're holding onto it, concurrent feature requests for the same solution
+            /// checksum can share the computation of that particular solution and avoid duplicated concurrent work.
+            /// </summary>
+            private readonly Dictionary<Checksum, SolutionAndInFlightCount> _solutionChecksumToSolution = new();
+
+            public ChecksumToSolutionCache(SemaphoreSlim gate)
+            {
+                _gate = gate;
+            }
+
+            public async ValueTask<SolutionAndInFlightCount?> TryFastGetSolutionAndAddInFlightCountAsync(
+                Checksum solutionChecksum,
+                CancellationToken cancellationToken)
+            {
+                using (await _gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+                {
+                    // From this point on we are mutating state.  Ensure we absolutely do not cancel accidentally.
+                    cancellationToken = default;
```

Comment:
once we acquire these locks, i always null-out the cancellation token so we don't even write an errant line that causes us to cancel after we've done some mutation.